pkg/comp-functions/functions/vshnnextcloud/deploy.go

```diff
@@ -0,0 +1,395 @@
+)
+
+//go:embed files/000-default.conf
+var apacheVhostConfig string
+
+//go:embed files/vshn-nextcloud.config.php
+var nextcloudConfig string
+
+// DeployNextcloud deploys a nexctloud instance via the codecentric Helm Chart.
+func DeployNextcloud(ctx context.Context, svc *runtime.ServiceRuntime) *xfnproto.Result {
+
+	comp := &vshnv1.VSHNNextcloud{}
+	err := svc.GetObservedComposite(comp)
+	if err != nil {
+		return runtime.NewFatalResult(fmt.Errorf("cannot get composite: %w", err))
+	}
+
+	svc.Log.Info("Bootstrapping instance namespace and rbac rules")
+	err = common.BootstrapInstanceNs(ctx, comp, comp.GetServiceName(), comp.GetName()+"-instanceNs", svc)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Sprintf("cannot bootstrap instance namespace: %s", err))
+	}
+
+	if !comp.Spec.Parameters.Service.DefaultInternalDB {
+		svc.Log.Info("Adding postgresql instance")
+		err = addPostgreSQL(svc, comp)
+		if err != nil {
+			return runtime.NewWarningResult(fmt.Sprintf("cannot create postgresql instance: %s", err))
+		}
+
+		svc.Log.Info("Checking readiness of cluster")
+
+		resourceCDMap := map[string][]string{
+			comp.GetName() + pgInstanceNameSuffix: {
+				vshnpostgres.PostgresqlHost,
+				vshnpostgres.PostgresqlPort,
+				vshnpostgres.PostgresqlDb,
+				vshnpostgres.PostgresqlUser,
+				vshnpostgres.PostgresqlPassword,
+			},
+		}
+
+		ready, err := svc.WaitForDependenciesWithConnectionDetails(comp.GetName(), resourceCDMap)
+		if err != nil {
+			// We're returning a fatal here, so in case something is wrong we won't delete anything by mistake.
+			return runtime.NewFatalResult(err)
+		} else if !ready {
+			return runtime.NewWarningResult("postgresql instance not yet ready")
+		}
+	}
+
+	svc.Log.Info("Adding release")
+
+	adminSecret, err := common.AddCredentialsSecret(comp, svc, []string{adminPWSecretField}, common.AddStaticFieldToSecret(map[string]string{adminUserSecretField: "admin"}))
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Sprintf("cannot generate admin secret: %s", err))
+	}
+
+	cd, err := svc.GetObservedComposedResourceConnectionDetails(adminSecret)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Sprintf("cannot get observed connection details for nextcloud admin: %s", err))
+	}
+
+	svc.SetConnectionDetail(adminPWConnectionDetailsField, cd[adminPWSecretField])
+	svc.SetConnectionDetail(adminConnectionDetailsField, cd[adminUserSecretField])
+	svc.SetConnectionDetail(hostConnectionDetailsField, []byte(fmt.Sprintf("%s-%s.%s.svc.cluster.local", comp.GetName(), serviceSuffix, comp.GetInstanceNamespace())))
+
+	err = addApacheConfig(svc, comp)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Sprintf("cannot add configmap for apache: %s", err))
+	}
+
+	err = addRelease(ctx, svc, comp, adminSecret)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Sprintf("cannot create release: %s", err))
+	}
+
+	return nil
+}
+
+func addPostgreSQL(svc *runtime.ServiceRuntime, comp *vshnv1.VSHNNextcloud) error {
```

Comment:
This looks very much copy pasted from Keycloak. Maybe something to make generic?

Reply:
I thought the same. But it's gonna make it rather complicated as I can't use our interface for the composition, since I need to access various fields in the composition. We should quickly discuss how to do that properly.

Reply:
>  I can't use our interface for the composition

You can add functions to it as needed. Or maybe even specify a new interface specifically for composites with embedded PostgreSQLs.

Reply:
I've refactored the function and moved it to the `common` package